design.md

```diff
@@ -1,2 +1,97 @@
 # TCP Load Balancer Design
 
+## Terminology/Assumptions
+
+- **Upstream**: A destination for TCP traffic. Could be an application running on HTTP, a gRPC server, or any other protocols using HTTP.
+- **Host**: An individual TCP address a client would connect to to reach an upstream.
+
+For example, upstream A is a REST server running on three machines. We would say upstream A has 3 hosts. A client would connect to the load balancer to try to reach upstream A, and the load balancer would appropriately route the client to one of the 3 hosts. (Assuming the client authenticated, passed authorization checks, etc.)
+
+## Libraries
+
+### Request forwarder
+
+The request forwarding library will implement a least-connections load balancing algorithm to forward TCP connections to a collection of known hosts for a given upstream. A library client (like our server) would create a load balancer for a given upstream using the following signature:
+
+```go
+func NewLoadBalancer(hosts []*net.TCPAddr) *LoadBalancer
+```
+
+When created, the load balancer would create an internal map of each host to the number of connections the host is currently handling.
+
+When the server gets a TCP connection, it will pass it to the request forwarding library to route it to the appropriate upstream:
+
+```go
+func (lb *LoadBalancer) HandleConnection(ctx context.Context, conn net.Conn)
+```
+
+At a high level, the `HandleConnection` function will:
+
+- Iterate over each of the hosts to find the one with the fewest connections
+    - The counters used for connections will be atomic (e.g. protected from concurrent access with a mutex, or use `sync/atomic` functions) to avoid race conditions
+- Dial the host to initiate a TCP connection with it
+    - If the connection can't be established, write an error message to the `conn` passed in. Close the connection and return without further processing.
+    - To avoid deadlocks, we'll use a timeout here to bail if the host doesn't respond in t seconds.
+- Increment (again, thread safely) the connection count for the host
+- Defer decrementing the connection count for the host
+- In two separate goroutines, `io.Copy` the connection passed in to the connection to the upstream and vice versa
+- Block for each goroutine to signal its completion
+- After completion, close both connections
+
+### Client Rate Limiter
+
+This library will implement a simple, in-memory rate limiter for client requests. Users will create a rate limiter by passing in a map of client identifier -> # of requests allowed per second like so:
```

Comment:
If it would be simpler, I think it's fine to use the same number of requests per second for all clients rather than have separate per-client limits. Up to you.

Reply:
Thanks - I think it'd be a _bit_ simpler but not too much; this design would just require creating a couple rate limiters in the server instead of one overall one. I may take you up on the offer if that proves more challenging than I thought though :laughing: